creator/referral_tokens/schema.py

```diff
@@ -0,0 +1,226 @@
+        " exchanging the token",
+    )
+
+
+class CreateReferralTokenMutation(graphene.Mutation):
+    class Arguments:
+        input = ReferralTokenInput(
+            required=True, description="Attributes for the new referral token"
+        )
+
+    referral_token = Field(ReferralTokenNode)
+
+    def mutate(self, info, input):
+        """
+        Create a new referral token
+        """
+
+        user = info.context.user
+
+        if not user.has_perm("referral_tokens.add_referraltoken"):
+            raise GraphQLError("Not allowed")
+
+        studies = get_studies(input)
+        groups = get_groups(input)
+
+        existing_token = (
+            ReferralToken.objects.filter(email=input["email"])
+            .filter(
+                created_at__lte=datetime.now()
+                + timedelta(days=settings.REFERRAL_TOKEN_EXPIRATION_DAYS)
+            )
+            .count()
+        )
+
+        if existing_token > 0:
+            raise GraphQLError("Invite alreay sent, awaiting user response")
+
+        referral_token = ReferralToken(email=input["email"],)
+        referral_token.save()
+        referral_token.studies.set(studies)
+        referral_token.groups.set(groups)
+        referral_token.save()
+
+        # Log an event
+        if user:
+            message = (
+                f"{user.username} invited {referral_token.email} to"
+                " {len(input['studies'])} studies"
+            )
+        else:
+            message = (
+                f"{referral_token.email} was invited to"
+                " {len(input['studies'])} studies"
+            )
+
+        event = Event(description=message, event_type="RT_CRE")
+        if not user._state.adding:
+            event.user = user
+        event.save()
+
+        return CreateReferralTokenMutation(referral_token=referral_token)
+
+
+class ExchangeReferralTokenMutation(graphene.Mutation):
+    class Arguments:
+        token = graphene.ID(
+            required=True, description="The id of the referral token"
+        )
+
+    message = graphene.String()
+    user = graphene.Field(UserNode)
+
+    def mutate(self, info, token):
+        """
+        Exchange a referral token
+        """
+
+        user = info.context.user
+
+        if not user.is_authenticated:
+            raise GraphQLError("Not allowed")
+
+        try:
+            _, uuid = from_global_id(token)
+            referral_token = ReferralToken.objects.get(uuid=uuid)
+        except ReferralToken.DoesNotExist:
+            raise GraphQLError("Referral token does not exist.")
+
+        if not referral_token.is_valid:
+            raise GraphQLError("Referral token is not valid.")
+
+        user.studies.set(referral_token.studies.all())
+        user.groups.set(referral_token.groups.all())
+        referral_token.claimed = True
+        referral_token.save()
+
+        message = "Studies are added to this user"
+
+        return ExchangeReferralTokenMutation(message=message, user=user)
```

Comment:
Maybe we should also return the studies and groups that were in the token? I'm not sure if we need the message because it only ever returns one value. We know it fails if we get back errors and can assume it works otherwise.